numpy/core/einsumfunc.pyi

```diff
@@ -0,0 +1,138 @@
+import sys
+from typing import List, TypeVar, Optional, Any, overload, Union, Tuple, Sequence
+
+from numpy import (
+    ndarray,
+    dtype,
+    bool_,
+    unsignedinteger,
+    signedinteger,
+    floating,
+    complexfloating,
+    number,
+    _OrderKACF,
+)
+from numpy.typing import (
+    _ArrayOrScalar,
+    _ArrayLikeBool_co,
+    _ArrayLikeUInt_co,
+    _ArrayLikeInt_co,
+    _ArrayLikeFloat_co,
+    _ArrayLikeComplex_co,
+    _DTypeLikeBool,
+    _DTypeLikeUInt,
+    _DTypeLikeInt,
+    _DTypeLikeFloat,
+    _DTypeLikeComplex,
+    _DTypeLikeComplex_co,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+_ArrayType = TypeVar(
+    "_ArrayType",
+    bound=ndarray[Any, dtype[Union[bool_, number[Any]]]],
+)
```

Comment:
I've noticed a bit of an inconsistency in how `np.einsum` handles the `out` parameter compared to the ufuncs. 
Namely, the former will _only_ accept boolean- or number-based arrays while the latter is more flexible in that sense, accepting the likes of string- or object-arrays without issue.

Example
---------
``` python
In [1]: import numpy as np

In [2]: AR_i = np.arange(3)
   ...: AR_U = np.empty(3, dtype='U23')

In [3]: np.einsum('i,i->i', AR_i, AR_i, out=AR_U)
---------------------------------------------------------------------------
SystemError: <built-in function c_einsum> returned a result with an error set

In [4]: np.multiply(AR_i, AR_i, out=AR_U)
Out[4]: array(['0', '1', '4'], dtype='<U23')
```